File: week05/3/prob3/middle_node.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
        
def middleNode(self, head):
    
    # initialization: two pointers
    slow = head # 1 step
    fast = head # 2 steps

    # traversal
    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next
    
    return slow

'''
In an even-length list, the fast pointer reaches the end
after an odd number of iterations (i.e., it steps through pairs of nodes).
The moment the fast pointer cannot move forward by two steps
(i.e., when it reaches the end or null),
the slow pointer has just passed the first middle node and is now on the second middle node.
'''

File: week05/3/prob3/middle_node.c
#include <stdlib.h>

/**
 * Definition for singly-linked list.
 */
struct ListNode {
    int val;
    struct ListNode *next;
};

struct ListNode* middleNode(struct ListNode* head) {
    
    // initialization: two pointers
    struct ListNode* slow = head; // 1 step
    struct ListNode* fast = head; // 2 steps

    // traversal
    while (fast != NULL && fast->next != NULL) {
        slow = slow->next;        // slow pointer moves 1 step
        fast = fast->next->next;  // fast pointer moves 2 steps
    }

    return slow; // slow pointer will now point to the middle node
}
